src/hotspot/share/opto/superword.cpp

```diff
@@ -1562,90 +1569,135 @@ void SuperWord::combine_packs() {
-          print_pack(p1);
-        }
+  if (is_trace_superword_packset()) {
+    tty->print_cr("\nAfter Superword::combine_packs");
+    print_packset();
+  }
 #endif
-        _packset.at_put(i, nullptr);
-        continue;
-      }
-      if (psize > max_vlen) {
-        Node_List* pack = new Node_List();
-        for (uint j = 0; j < psize; j++) {
-          pack->push(p1->at(j));
-          if (pack->size() >= max_vlen) {
-            assert(is_power_of_2(pack->size()), "sanity");
-            _packset.append(pack);
-            pack = new Node_List();
-          }
-        }
-        _packset.at_put(i, nullptr);
+}
+
+void SuperWord::split_packs_for_max_vector_size() {
+  assert(!_packset.is_empty(), "packset not empty");
+
+  for (int i = 0; i < _packset.length(); i++) {
+    Node_List* pack = _packset.at(i);
+    assert(pack != nullptr, "no nullptr in packset");
+    uint max_vlen = max_vector_size_in_def_use_chain(pack->at(0));
+    assert(is_power_of_2(max_vlen), "sanity");
+    uint pack_size = pack->size();
+    if (pack_size <= max_vlen) {
+      continue;
+    }
+    // Split off the "upper" nodes into new packs
+    Node_List* new_pack = new Node_List();
+    for (uint j = max_vlen; j < pack_size; j++) {
+      Node* n = pack->at(j);
+      // is new_pack full?
+      if (new_pack->size() >= max_vlen) {
+        assert(is_power_of_2(new_pack->size()), "sanity");
+        _packset.append(new_pack);
+        Node_List* new_pack = new Node_List();
       }
+      new_pack->push(n);
+    }
+    // remaining new_pack
+    if (new_pack->size() > 1) {
+      _packset.append(new_pack);
+    }
+    // truncate
+    while (pack->size() > max_vlen) {
+      pack->pop();
     }
   }
 
-  // We know that the nodes in a pair pack were independent - this gives us independence
-  // at distance 1. But now that we may have more than 2 nodes in a pack, we need to check
-  // if they are all mutually independent. If there is a dependence we remove the pack.
-  // This is better than giving up completely - we can have partial vectorization if some
-  // are rejected and others still accepted.
-  //
-  // Examples with dependence at distance 1 (pack pairs are not created):
-  // for (int i ...) { v[i + 1] = v[i] + 5; }
-  // for (int i ...) { v[i] = v[i - 1] + 5; }
-  //
-  // Example with independence at distance 1, but dependence at distance 2 (pack pairs are
-  // created and we need to filter them out now):
-  // for (int i ...) { v[i + 2] = v[i] + 5; }
-  // for (int i ...) { v[i] = v[i - 2] + 5; }
-  //
-  // Note: dependencies are created when a later load may reference the same memory location
-  // as an earlier store. This happens in "read backward" or "store forward" cases. On the
-  // other hand, "read forward" or "store backward" cases do not have such dependencies:
-  // for (int i ...) { v[i] = v[i + 1] + 5; }
-  // for (int i ...) { v[i - 1] = v[i] + 5; }
```

Comment:
Note: I moved this to `filter_packs_for_mutual_independence`, and changed indentation.